File: services/TalentBridge.Post.Api/JobPostings/JobPostingValidators.cs
using FluentValidation;
using TalentBridge.Post.Api.Validation;
using TalentBridge.Post.Domain.JobPostings;

namespace TalentBridge.Post.Api.JobPostings;

/// <summary>
/// The authoritative rules from CLAUDE.md §6. The Angular client mirrors them
/// for latency only. Property names become the camelCase keys the client maps
/// onto its form controls, so cross-field failures are attached to the control
/// the user should fix (salary range → salaryMax).
/// </summary>
public class CreateJobPostingValidator : AbstractValidator<CreateJobPostingRequest>
{
    public const decimal MaxSalary = 10_000_000m;

    public CreateJobPostingValidator(TimeProvider clock)
    {
        RuleFor(x => x.Title)
            .NotEmpty().WithMessage("Job title is required.")
            .Length(3, 120).WithMessage("Job title must be 3–120 characters.");

        RuleFor(x => x.Department)
            .NotEmpty().WithMessage("Department is required.")
            .Length(2, 80).WithMessage("Department must be 2–80 characters.");

        RuleFor(x => x.EmploymentType)
            .NotEmpty().WithMessage("Employment type is required.")
            .Must(BeOneOf(JobPostingVocabulary.EmploymentTypes)).WithMessage(OneOfMessage("Employment type", JobPostingVocabulary.EmploymentTypes));

        RuleFor(x => x.Seniority)
            .NotEmpty().WithMessage("Seniority is required.")
            .Must(BeOneOf(JobPostingVocabulary.Seniorities)).WithMessage(OneOfMessage("Seniority", JobPostingVocabulary.Seniorities));

        RuleFor(x => x.Openings)
            .NotNull().WithMessage("Number of openings is required.")
            .InclusiveBetween(1, 999).WithMessage("Number of openings must be a whole number from 1 to 999.");

        RuleFor(x => x.WorkArrangement)
            .NotEmpty().WithMessage("Work arrangement is required.")
            .Must(BeOneOf(JobPostingVocabulary.WorkArrangements)).WithMessage(OneOfMessage("Work arrangement", JobPostingVocabulary.WorkArrangements));

        // Location is required unless the role is Remote, where the country stands in.
        RuleFor(x => x.Location)
            .NotEmpty().WithMessage("Location is required unless the work arrangement is Remote.")
            .When(x => x.WorkArrangement != "Remote");
        RuleFor(x => x.Location)
            .Length(2, 120).WithMessage("Location must be 2–120 characters.")
            .When(x => !string.IsNullOrWhiteSpace(x.Location));

        RuleFor(x => x.Country)
            .NotEmpty().WithMessage("Country is required.")
            .Length(2, 80).WithMessage("Country must be 2–80 characters.");

        RuleFor(x => x.SalaryMin)
            .NotNull().WithMessage("Salary minimum is required.")
            .GreaterThan(0).WithMessage("Salary minimum must be greater than 0.")
            .LessThanOrEqualTo(MaxSalary).WithMessage("Salary minimum must be at most 10,000,000.");

        RuleFor(x => x.SalaryMax)
            .NotNull().WithMessage("Salary maximum is required.")
            .GreaterThan(0).WithMessage("Salary maximum must be greater than 0.")
            .LessThanOrEqualTo(MaxSalary).WithMessage("Salary maximum must be at most 10,000,000.");

        // Cross-field rule, keyed to salaryMax by contract (§6). Only evaluated
        // once both values are individually valid so the user sees one error.
        RuleFor(x => x.SalaryMax)
            .GreaterThan(x => x.SalaryMin).WithMessage("Salary maximum must be greater than salary minimum.")
            .When(x => x.SalaryMin is > 0 and <= MaxSalary && x.SalaryMax is > 0 and <= MaxSalary);

        RuleFor(x => x.SalaryCurrency)
            .NotEmpty().WithMessage("Currency is required.")
            .Must(c => c is not null && JobPostingVocabulary.Currencies.Contains(c.ToUpperInvariant()))
                .WithMessage(OneOfMessage("Currency", JobPostingVocabulary.Currencies));

        RuleFor(x => x.PayPeriod)
            .NotEmpty().WithMessage("Pay period is required.")
            .Must(BeOneOf(JobPostingVocabulary.PayPeriods)).WithMessage(OneOfMessage("Pay period", JobPostingVocabulary.PayPeriods));

        RuleFor(x => x.Description)
            .NotEmpty().WithMessage("Job description is required.")
            .Length(50, 10_000).WithMessage("Job description must be 50–10,000 characters.");

        RuleFor(x => x.Responsibilities).MaximumLength(10_000).WithMessage("Responsibilities must be at most 10,000 characters.");
        RuleFor(x => x.Requirements).MaximumLength(10_000).WithMessage("Requirements must be at most 10,000 characters.");

        // The chip input is a single form control, so both rules key to "skills".
        RuleFor(x => x.Skills)
            .Must(s => s is null || s.Length <= JobPosting.MaxSkills).WithMessage("Add at most 20 skills.")
            .Must(s => s is null || s.All(e => e is not null && e.Trim().Length is >= 1 and <= 40))
                .WithMessage("Each skill must be 1–40 characters.")
                .When(x => x.Skills is { Length: <= JobPosting.MaxSkills }, ApplyConditionTo.CurrentValidator);

        RuleFor(x => x.ApplicationEmail)
            .MustBeEmailAddress().WithMessage("Enter a valid application email address.")
            .When(x => !string.IsNullOrWhiteSpace(x.ApplicationEmail));

        RuleFor(x => x.ApplicationUrl)
            .Must(BeAbsoluteHttpUrl).WithMessage("Application URL must be an absolute http:// or https:// link.")
            .MaximumLength(2048)
            .When(x => !string.IsNullOrWhiteSpace(x.ApplicationUrl));

        RuleFor(x => x.ReferenceCode)
            .MaximumLength(40).WithMessage("Internal reference code must be at most 40 characters.");

        RuleFor(x => x.ClosingDate)
            .NotNull().WithMessage("Closing date is required.")
            .Must(d => d > DateOnly.FromDateTime(clock.GetUtcNow().UtcDateTime))
                .WithMessage("Closing date must be in the future.");

        RuleFor(x => x.Status)
            .NotEmpty().WithMessage("Status is required.")
            .Must(s => s is JobPostingStatus.Draft or JobPostingStatus.Published)
                .WithMessage("Status must be Draft or Published.");
    }

    private static Func<string?, bool> BeOneOf(IReadOnlyList<string> allowed) =>
        value => value is not null && allowed.Contains(value, StringComparer.Ordinal);

    private static string OneOfMessage(string field, IReadOnlyList<string> allowed) =>
        $"{field} must be one of: {string.Join(", ", allowed)}.";

    private static bool BeAbsoluteHttpUrl(string? url) =>
        Uri.TryCreate(url, UriKind.Absolute, out var uri)
        && (uri.Scheme == Uri.UriSchemeHttp || uri.Scheme == Uri.UriSchemeHttps)
        && !string.IsNullOrEmpty(uri.Host);
}

public sealed class UpdateJobPostingValidator : AbstractValidator<UpdateJobPostingRequest>
{
    public UpdateJobPostingValidator(TimeProvider clock)
    {
        Include(new CreateJobPostingValidator(clock));

        RuleFor(x => x.Version)
            .NotNull().WithMessage("Version is required — send the version you loaded.")
            .GreaterThanOrEqualTo(1).WithMessage("Version must be at least 1.");
    }
}

File: apps/post-web/src/app/features/postings/job-posting-validators.ts
import { AbstractControl, ValidationErrors, ValidatorFn } from '@angular/forms';

/**
 * Client mirrors of CLAUDE.md §6. They exist for latency only; the server is
 * the authority and its messages win when they arrive. Keys are the same the
 * server uses so a template can render either source the same way.
 */
export const MAX_SALARY = 10_000_000;
export const MAX_SKILLS = 20;

export function todayUtc(): string {
  return new Date().toISOString().slice(0, 10);
}

/** yyyy-mm-dd strictly after today's UTC date. */
export const futureDate: ValidatorFn = (control) => {
  const value = control.value as string | null;
  if (!value) return null;
  if (!/^\d{4}-\d{2}-\d{2}$/.test(value) || Number.isNaN(Date.parse(value))) return { dateFormat: true };
  return value > todayUtc() ? null : { future: true };
};

/** Group-level: min strictly below max, only once both are individually valid. Surfaced against salaryMax. */
export const salaryRange: ValidatorFn = (group) => {
  const min = group.get('salaryMin')?.value as number | null;
  const max = group.get('salaryMax')?.value as number | null;
  const valid = (n: number | null): n is number => typeof n === 'number' && n > 0 && n <= MAX_SALARY;
  return valid(min) && valid(max) && min >= max ? { salaryRange: 'Salary maximum must be greater than salary minimum.' } : null;
};

/** Location is required unless the arrangement is Remote (then the country stands in). */
export const locationUnlessRemote: ValidatorFn = (control) => {
  const arrangement = control.parent?.get('workArrangement')?.value as string | undefined;
  const value = ((control.value as string | null) ?? '').trim();
  if (arrangement !== 'Remote' && value.length === 0) return { required: true };
  if (value.length > 0 && (value.length < 2 || value.length > 120)) return { length: { min: 2, max: 120 } };
  return null;
};

export const skillsList: ValidatorFn = (control) => {
  const skills = (control.value as string[] | null) ?? [];
  if (skills.length > MAX_SKILLS) return { maxSkills: true };
  if (skills.some((s) => s.trim().length < 1 || s.trim().length > 40)) return { skillLength: true };
  return null;
};

export const absoluteHttpUrl: ValidatorFn = (control) => {
  const value = ((control.value as string | null) ?? '').trim();
  if (!value) return null;
  try {
    const url = new URL(value);
    return (url.protocol === 'http:' || url.protocol === 'https:') && url.hostname ? null : { url: true };
  } catch {
    return { url: true };
  }
};

export const emailAddress: ValidatorFn = (control) => {
  const value = ((control.value as string | null) ?? '').trim();
  if (!value) return null;
  return /^[^@\s]+@[^@\s]+\.[^@\s.]+$/.test(value) && value.length <= 320 ? null : { email: true };
};

export const wholeNumber: ValidatorFn = (control) => {
  const value = control.value as number | null;
  return value === null || Number.isInteger(value) ? null : { integer: true };
};

/** Human message for whichever error a control (or the group, for salaryMax) carries. */
export function messageFor(name: string, errors: ValidationErrors | null): string | null {
  if (!errors) return null;
  if (typeof errors['server'] === 'string') return errors['server'];
  if (typeof errors['salaryRange'] === 'string') return errors['salaryRange'];
  const label = FIELD_LABELS[name] ?? name;
  if (errors['required']) return `${label} is required.`;
  if (errors['minlength']) return `${label} must be at least ${errors['minlength'].requiredLength} characters.`;
  if (errors['maxlength']) return `${label} must be at most ${errors['maxlength'].requiredLength.toLocaleString()} characters.`;
  if (errors['length']) return `${label} must be ${errors['length'].min}–${errors['length'].max} characters.`;
  if (errors['min']) return `${label} must be greater than ${errors['min'].min - 1 < 0 ? 0 : errors['min'].min - 1}.`;
  if (errors['max']) return `${label} must be at most ${errors['max'].max.toLocaleString()}.`;
  if (errors['integer']) return `${label} must be a whole number.`;
  if (errors['future']) return 'Closing date must be in the future.';
  if (errors['dateFormat']) return 'Enter the closing date as yyyy-mm-dd.';
  if (errors['maxSkills']) return `Add at most ${MAX_SKILLS} skills.`;
  if (errors['skillLength']) return 'Each skill must be 1–40 characters.';
  if (errors['url']) return 'Application URL must be an absolute http:// or https:// link.';
  if (errors['email']) return 'Enter a valid application email address.';
  return 'Enter a valid value.';
}

export const FIELD_LABELS: Record<string, string> = {
  referenceCode: 'Internal reference code',
  title: 'Job title',
  department: 'Department',
  employmentType: 'Employment type',
  seniority: 'Seniority',
  openings: 'Number of openings',
  workArrangement: 'Work arrangement',
  location: 'Location',
  country: 'Country',
  salaryMin: 'Salary min',
  salaryMax: 'Salary max',
  salaryCurrency: 'Currency',
  payPeriod: 'Pay period',
  salaryVisible: 'Salary visibility',
  description: 'Job description',
  responsibilities: 'Responsibilities',
  requirements: 'Requirements',
  skills: 'Skills',
  applicationUrl: 'Application URL',
  applicationEmail: 'Application email',
  closingDate: 'Closing date',
  status: 'Status',
  version: 'Version',
};
